mailpoet/lib/Subscribers/EngagementDataBackfiller.php

```diff
@@ -0,0 +1,200 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Subscribers;
+
+use MailPoet\Entities\StatisticsClickEntity;
+use MailPoet\Entities\StatisticsNewsletterEntity;
+use MailPoet\Entities\StatisticsOpenEntity;
+use MailPoet\Entities\SubscriberEntity;
+use MailPoet\Segments\DynamicSegments\Filters\FilterHelper;
+use MailPoet\Segments\DynamicSegments\Filters\WooFilterHelper;
+use MailPoet\WooCommerce\Helper;
+use MailPoetVendor\Carbon\Carbon;
+use MailPoetVendor\Doctrine\DBAL\Connection;
+use MailPoetVendor\Doctrine\DBAL\ForwardCompatibility\Result;
+use MailPoetVendor\Doctrine\ORM\EntityManager;
+
+class EngagementDataBackfiller {
+  /** @var EntityManager */
+  private $entityManager;
+
+  /** @var Helper */
+  private $wcHelper;
+
+  /** @var WooFilterHelper */
+  private $wooFilterHelper;
+
+  /** @var FilterHelper */
+  private $filterHelper;
+
+  /** @var int */
+  private $lastProcessedSubscriberId = 0;
+
+  public function __construct(
+    EntityManager $entityManager,
+    WooFilterHelper $wooFilterHelper,
+    FilterHelper $filterHelper,
+    Helper $wcHelper
+  ) {
+    $this->entityManager = $entityManager;
+    $this->wcHelper = $wcHelper;
+    $this->wooFilterHelper = $wooFilterHelper;
+    $this->filterHelper = $filterHelper;
+  }
+
+  /**
+   * @return SubscriberEntity[]
+   */
+  public function getBatch(int $lastProcessedId = 0, int $batchSize = 100): array {
+    $subscribers = $this->entityManager->createQueryBuilder()
+      ->select('PARTIAL s.{id, email, lastPurchaseAt, lastClickAt, lastOpenAt, lastSendingAt}')
+      ->from(SubscriberEntity::class, 's')
+      ->where('s.id > :lastProcessedId')
+      ->setMaxResults($batchSize)
```

Comment:
Does it make sense to also order by `s.id`? Or can we rely on "by default it is always ordered by id?

Reply:
Thanks for catching this! I'll include an explicit orderBy.